### PR TITLE
Fix crash on duplicate submissionId

### PR DESCRIPTION
**Pull request by mziolekda**
- Fix crash in ConfigManagement- and PackageManagement- services on duplicate submissionIds from different participants.
- Extend ledger-api-test-tool to cover the duplicate submissionId cases.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (tudor-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:191]:
```suggestion
    "Changing config twice with the same submissionId succeeds on different participants",
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (tudor-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:230]: The description is the same as for the previous test. 
Looking better at them, they look highly similar. Did you consider only writing one test with the following order that would assert both cases:
```
_ <- alpha.setTimeModel(good_1)
failure <- alpha.setTimeModed(bad)
_ <- alpha.setTimeModel(good_2)
} yield ...
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (tudor-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:254]: A tad shorter and a bit more idiomatic
```suggestion
      oldTimeModel = response.timeModel.getOrElse(
        throw new AssertionError("Expected time model to be defined")
      )
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/PackageManagementServiceIT.scala
Comment (tudor-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/PackageManagementServiceIT.scala:86]: The same comment as for the `ConfigManagementServiceIT` wrt to test descriptions and similarity.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:111]: Is it a problem that a user can see a modified `submissionId` (through `ApiConfigManagementService.entries`) in comparison to what he has submitted with?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (mziolekda) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:230]: Done

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/PackageManagementServiceIT.scala
Comment (mziolekda) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/PackageManagementServiceIT.scala:86]: Done

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:111]: User will not see it through the Ledger API and within the server we have full autonomy.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (mziolekda) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:254]: Done

**Review by tudor-da — APPROVED** [review]
LGTM once the changes land in.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:211]: soo good!

**Review by nmarton-da — APPROVED** [review]
LGTM
was nice to see that removing the constraint made quite some stuff smaller, simpler

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:204]: Does this work on distributed ledgers?

The first line only returns after the config has been persisted on the ledger (because the first participant waits for the config change event).
However, it might take some time until index of the second participant has caught up with the new config. If it hasn't caught up, the check in `ApiConfigManagementService.setTimeModel` will fail.

The situation is complicated because we verify the config generation both in the participant server (`ApiConfigManagementService`, which relies on information from the index that is not up-to-date) and on the ledger (which should rely on the correct information).

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:267]: This relies on the error description. This is ok if we know the error comes from the participant, but we have no control over the description if it comes from the ledger (see previous comment).

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoConfigurationMutatingSpec.scala
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoConfigurationMutatingSpec.scala:11]:
```suggestion
// TODO append-only: remove this class once the mutating schema is removed
trait JdbcLedgerDaoConfigurationMutatingSpec {
```

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:227]: We have a very similar block in `storePartyEntry`. Should we remove it as well, or at least create an issue to track it?

With this PR, both the config management service and the party management service will send globally unique submission IDs to the ledger, so they should be treated the same in `JdbcLedgerDao`.

**Review by rautenrieth-da — COMMENTED** [review]
LGTM in general. I left a few comments regarding the tests, I think they might be flaky on some distributed ledgers.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (mziolekda) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:204]: Very good point! I will synchronize in between

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala:215]: You know `+` also works with strings, right? :upside_down_face:

I'd go with the interpolation.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:184]: Can you change `def uploadDarFile(ByteString)` to delegate to this one please?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:710]: Likewise, can you change the above `def setTimeModel` to use this one?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:47]: We already import `SubmissionId` directly.

```suggestion
    submissionIdGenerator: String => SubmissionId,
```

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:143]:
```suggestion
      submissionIdGenerator: String => SubmissionId = augmentSubmissionId,
```

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:159]:
```suggestion
  private def augmentSubmissionId: String => SubmissionId =
```

Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:159]: This seems to be almost the same as the one in `ApiConfigManagementService`, and the `isEmpty` check seems benign. Can I recommend factoring it out into a shared location?

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoSuite.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoSuite.scala:888]: Perhaps use `.andThen` here to avoid having to return the input.

```suggestion
      .andThen { case Success(_) =>
        previousOffset.set(Some(offset))
      }
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (mziolekda) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:221]: yes, this test is inherently racy on KV ledgers.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala
Comment (mziolekda) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ConfigManagementServiceIT.scala:204]: Done

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/PackageManagementServiceIT.scala
Comment (mziolekda) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/PackageManagementServiceIT.scala:72]: I wrote this test while I was attempting to generate an on-ledger error. This indeed is not possible with the current level of pre-execution validation.
There is a slim potential that in the future this validation is relaxed and we might end-up in a bug. I agree that this is probably too far fetched. I will remove those tests.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoConfigurationMutatingSpec.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoConfigurationMutatingSpec.scala:11]: Done

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/PackageManagementServiceIT.scala
Comment (mziolekda) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/PackageManagementServiceIT.scala:72]: Done

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiConfigManagementService.scala:215]: Done

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:159]: Done

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (mziolekda) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:184]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/JdbcLedgerDao.scala:227]: You are absolutely right, this should be changed as well. I will create a ticket:
https://digitalasset.atlassian.net/browse/DPP-504

**Pull request merged**
